### PR TITLE
Basic sliding window

### PR DESCRIPTION
**Pull request by kexinrong**
- Implemented naive window aggregate function classes: BatchedWindowMax, BatchedWindowCount, BatchedWindowSum
- Implemented simple sliding window transform that outputs one datum for each pane.

### CODE REVIEW

File: src/main/java/macrobase/analysis/transform/SlidingWindowTransform.java [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+import macrobase.datamodel.Datum;
+
+import java.util.ArrayList;
+import java.util.List;
+
+
+public class SlidingWindowTransform extends FeatureTransform {
+    public int minPaneSize;
+    public BatchedWindowAggregate windowAggregate;
+    public int timeColumn;
```
Comment (pbailis) [inline src/main/java/macrobase/analysis/transform/SlidingWindowTransform.java]: Is there any reason these are public? I would expect `private final`

File: src/main/java/macrobase/analysis/transform/aggregate/BatchedWindowAggregate.java [hunk 2]
```diff
@@ -0,0 +1,12 @@
+package macrobase.analysis.transform.aggregate;
+
+import macrobase.datamodel.Datum;
+import java.util.List;
+
+public abstract class BatchedWindowAggregate implements IncrementalWindowAggregate {
```
Comment (pbailis) [inline src/main/java/macrobase/analysis/transform/aggregate/BatchedWindowAggregate.java]: What if instead this class implements `process(List<Datum> new_data, List<Datum> expired_data)` (from `IncrementalWindowAggregate`?

Basic idea: BatchedWindowAggregate can keep track of the current window, then, each time `process(new_data, expired_data)` is called, it updates the window and passes the entire window to an abstract `public Datum process(List<Datum> data)` function.

File: src/main/java/macrobase/analysis/transform/SlidingWindowTransform.java [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+import macrobase.conf.MacroBaseDefaults;
+import macrobase.datamodel.Datum;
+
+import java.util.ArrayList;
+import java.util.List;
+
+
+public class SlidingWindowTransform extends FeatureTransform {
+    public int minPaneSize;
+    public BatchedWindowAggregate windowAggregate;
```
Comment (pbailis) [inline src/main/java/macrobase/analysis/transform/SlidingWindowTransform.java]: Can this instead be `IncrementalWindowAggregate`?

File: src/main/java/macrobase/analysis/transform/aggregate/IncrementalWindowAggregate.java [hunk 4]
```diff
@@ -0,0 +1,8 @@
+package macrobase.analysis.transform.aggregate;
+
+import macrobase.datamodel.Datum;
+import java.util.List;
+
+public interface IncrementalWindowAggregate {
```
Comment (pbailis) [inline src/main/java/macrobase/analysis/transform/aggregate/IncrementalWindowAggregate.java]: Or just `WindowedAggregate`!

File: src/main/java/macrobase/analysis/transform/SlidingWindowTransform.java (hunk 3 above)
Comment (pbailis) [inline src/main/java/macrobase/analysis/transform/SlidingWindowTransform.java]: Put another way, if every aggregate implements `IncrementalWindowAggregate`, then all this function needs to do is keep track of panes and when they enter and expire...

File: src/test/java/macrobase/analysis/transform/SlidingWindowTransformTest.java [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+            Datum d = transformed.get(i);
+            assertTrue(d.getMetrics().getEntry(0) == i * 3);
+            if (i == 3)
+                assertTrue(d.getMetrics().getEntry(1) == 1);
+            else
+                assertTrue(d.getMetrics().getEntry(1) == 3);
+
+        }
+
+        /* Test a different aggregation function */
```
Comment (pbailis) [inline src/test/java/macrobase/analysis/transform/SlidingWindowTransformTest.java]: Minor: would be nice to have these in separate functions to help tell what part of the functionality being tested fails (if and when it fails)!

**Comment by kexinrong** [comment]
Main changes include:
- Change the WindowAggregate interface from _process(List<Datum> data)_ to _updateWindow(List<Datum> new_data, List<Datum> old_data)_
- Each window aggregate class can indicate whether they support panes. If so, the _updateWindow_ function takes in lists of panes; otherwise it takes in lists of datum.
- Refactored _SlidingWindowTransform_ so that it produces a stream of window aggregates. Incomplete window at the end of the data stream can be flushed out during shutdown. 

This ended up being more complicated than I thought. Happy to discuss the changes online/in person.

**Comment by pbailis** [comment]
@kexinrong LMK when you have an update to look at!

**Comment by kexinrong** [comment]
Please refer to PR #136.

**Pull request closed without merging**